### PR TITLE
UCT/DC: fixes dci leak of ep pending queue

### PR DESCRIPTION
**Pull request by alex-mikheev**
@yosefe please take a look

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3011/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/547/ for details.

File: src/uct/ib/dc/base/dc_ep.c [hunk 1]
```diff
@@ -32,6 +32,16 @@ static UCS_CLASS_CLEANUP_FUNC(uct_dc_ep_t)
     /* TODO: this is good for dcs policy only.
      * Need to change if eps share dci
      */
+
+    if (!uct_dc_iface_dci_has_outstanding(iface, self->dci)) {
+        ucs_fatal("ifface (%p) ep (%p) dci leak detected: dci=%d",
```
Comment (brminich) [inline src/uct/ib/dc/base/dc_ep.c]: Would not be ucs_error be enough here? Even though this is a faulty case, communication can still continue, right?

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2486/ for details (Mellanox internal link).

File: src/uct/ib/dc/base/dc_ep.c (hunk 1 above)
Comment (yosefe) [inline src/uct/ib/dc/base/dc_ep.c]: also a typo here: "ifface"

File: src/uct/ib/dc/base/dc_ep.c [hunk 2]
```diff
@@ -129,6 +139,16 @@ uct_dc_iface_dci_do_pending_tx(ucs_arbiter_t *arbiter,
...
                    ucs_status_string(status));
     if (status == UCS_OK) {
+        /* Release dci if this is the last elem in the group and the dci has no
+         * outstanding operations. For example pending callback did not send
+         * anything. (uct_ep_flush or just return ok) 
+         *
+         * note: arbiter removes elem _after_ dispatch, so a little hack is
+         * used
+         */ 
+        if (elem == ep->arb_group.tail) {
```
Comment (yosefe) [inline src/uct/ib/dc/base/dc_ep.c]: IMHO better to add function like ucs_arbiter_elem_is_last(group, elem)

File: src/uct/ib/dc/base/dc_ep.h [hunk 3]
```diff
@@ -168,6 +169,27 @@ static inline void uct_dc_iface_dci_alloc_dcs(uct_dc_iface_t *iface, uct_dc_ep_t
...
 }
 
+static inline void uct_dc_iface_dci_free_dcs(uct_dc_iface_t *iface, uct_dc_ep_t *ep)
+{
+    uint8_t dci;
+
+    ucs_assert(ep->dci != UCT_DC_EP_NO_DCI);
+    ucs_assert(iface->tx.stack_top > 0);
+
+    dci = ep->dci;
```
Comment (yosefe) [inline src/uct/ib/dc/base/dc_ep.h]: could be done in line 174

File: test/gtest/ucp/ucp_test.cc [hunk 4]
```diff
@@ -380,6 +385,7 @@ void* ucp_test_base::entity::disconnect_nb(int ep_index) const {
 }
 
 void ucp_test_base::entity::destroy_worker(int worker_index) {
+    flush_worker(worker_index);
```
Comment (yosefe) [inline test/gtest/ucp/ucp_test.cc]: why is that needed? flush_ep @ disconnect should be enough

File: test/gtest/uct/test_dc.cc [hunk 5]
```diff
@@ -46,10 +46,99 @@ class test_dc : public uct_test {
...
+    static ucs_log_func_rc_t
+    log_handler_func(const char *file, unsigned line, const char *function,
+                     ucs_log_level_t level, const char *prefix, const char *message,
+                     va_list ap)
+    {
+        if (level != UCS_LOG_LEVEL_WARN) {
+            /* debug messages are ignored */
+            return UCS_LOG_FUNC_RC_CONTINUE;
+        }
+        n_warnings++;
```
Comment (yosefe) [inline test/gtest/uct/test_dc.cc]: need to search for some specific words in the warning message

File: src/uct/ib/dc/base/dc_ep.c (hunk 1 above)
Comment (alex-mikheev) [inline src/uct/ib/dc/base/dc_ep.c]: If this happens it means that there is some kind of internal error in the dc tl code. It is best to stop right here to simplify debugging and because further usage of dc tl may lead to unpredictable results.

I can change fatal to abort to make things clearer.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3015/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/557/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2497/ for details (Mellanox internal link).

File: src/uct/ib/dc/base/dc_ep.c [hunk 6]
```diff
@@ -32,11 +32,9 @@ static UCS_CLASS_CLEANUP_FUNC(uct_dc_ep_t)
     /* TODO: this is good for dcs policy only.
      * Need to change if eps share dci
      */
-
-    if (!uct_dc_iface_dci_has_outstanding(iface, self->dci)) {
-        ucs_fatal("ifface (%p) ep (%p) dci leak detected: dci=%d",
-                  iface, self, self->dci);
-    }
+    ucs_assertv_always(uct_dc_iface_dci_has_outstanding(iface, self->dci) > 0,
```
Comment (yosefe) [inline src/uct/ib/dc/base/dc_ep.c]: minor - looks like no real need for the ">0"?

**Comment by yosefe** [comment]
bot:ornl:retest

**Comment by brminich** [comment]
:+1:

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3025/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/566/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3027/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2506/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2509/ for details (Mellanox internal link).
